Cure HTTP Authentication Singleton's Password Overwrite

jheinnic opened the pull request:
HTTP Basic Authentication support in the Schema Registry Client that Registry Helper relies on uses a Singleton pattern to implement its strategies each of the three credential sources it implements a strategy for interfacing with.

Because Registry Helper uses the client for both source as well as destination access, this causes a problem if a user configured both source and destination authentication, and uses different credentials, but acquires them from the same source strategy on both ends of the line.

At runtime, registry helper first initialized the source registry client with no issues, then it proceeds to configure the destination registry client and ends up accessing the same basic HTTP Authentication helper object already in use by the source registry.  It winds up overwriting the credentials and wires the resulting client to the destination registry.  The first time the SMT attempts to read a schema using the source registry's client object, it fails due to the fact that the credentials are now for the destination registry instead of its own.

This is a little complicated to fix--we cannot modify the client to replace the existing implementation with one that is not a singleton without breaking encapsulation.  For now, this PR only includes additional test cases, with the intent to discuss an approach to fix the underlying cause and add it to the PR before it is merged.

I think the most likely approach to work is going to involve either creating a second set of identical implementations of the three basic HTTP authentication singletons.  A prefix or a suffix would be added to the three new instances and the registry client would re-map the service's lookup key (USER_INFO, URL, or SASL) to include the appropriate prefix for either the source or destination client.

I'm not sure if the shade plugin could be configured in a way to assist with this issue.  It might be able to handle a second relocation, but I don't think it could be told to add a prefix or suffix to the services' lookup keys in the process.  The former approach might require making a hand copy of the three strategies as part of this project's own codebase.  I'm open to suggestions!

Alternately, we could pursue a more heavy weight but more conservative approach and arrange for a second relocation of the entire client, perhaps by using the shade plugin in a dependency to relocate the client twice, once to a namespace intended for the source, and again to a different namespace intended for the destination.

jheinnic commented:
This branch now has a solution to its associated complication that leverages "SRC" and "DEST" prefixes to give Schema Registry a distinct singleton for each direction of Schema Registry transfer, so no more overwritten credentials.

cricket007 commented:
Hey @jheinnic , would you mind making a new comment to update the README to include all these new properties? 

Happy to merge after that

jheinnic commented:
I seem to recall having thought the name spacing was reconciling some early issue I had with the shade plugin, but I can neither recall what that was nor reproduce it by staying within the `cricket.jmoore` namespace, so I've renamed these with `Src` and `Dest` prefixes and moved them to a closer package namespace.  I was successfully able to subclass the original concrete types with only the alias method override required to give them each a unique lookup key that does not conflict with the repository client's own original singleton instances.

No problem getting the new test case to pass as-is and a manual inspection of the jar file revealed no surprised.  I'm folding the jar into a container now as a last precaution, but expect no trouble!

Thanks for the sound feedback!

jheinnic commented:
I see this PR has been approved, but it looks like its merge is still pending, FYI.

cricket007 commented:
> I'm folding the jar into a container now as a last precaution, but expect no trouble!

Sounded like you were still building it, and were going to test out those changes? Or you think it's good to go?

jheinnic commented:
Coming back to this a little late, but I finally managed to get a copy of an image with this change included pre-merge and confirmed that it does indeed work as desired.  It was a little harder this time than with previous commits because I'm not longer adding a jar to the image, but attempting to override one and had to figure out a way to do that after the container launched, but before connect's isolating class loader saw the jar file.  I finally realized I could suspend the launch script without losing the ability to `docker cp` files in by setting environment flags so the launcher would suspend-on-startup waiting for a debugger to be attached.

Anyhow, I see you've already merged the request, so will go ahead and revert my workarounds from our deployment and go back to using a build of our base image that includes this latest commit.  Thanks again for taking the submission!
